specs/sharding/beacon-chain.md

```diff
@@ -668,119 +701,103 @@ def process_epoch(state: BeaconState) -> None:
-                if (c.slot, c.shard, c.confirmed) == (slot, shard, True)
-            ]
-            if not any(confirmed_candidates):
-                continue
-            candidate = confirmed_candidates[0]
-
-            # Charge EIP 1559 fee
-            proposer = get_shard_proposer_index(state, slot, shard)
-            fee = (
-                (state.shard_gasprice * candidate.commitment.length)
-                // TARGET_SAMPLES_PER_BLOCK
-            )
-            decrease_balance(state, proposer, fee)
-
-            # Track updated gas price
-            new_gasprice = compute_updated_gasprice(
-                new_gasprice,
-                candidate.commitment.length,
-                adjustment_quotient,
-            )
+        buffer_index = slot % SHARD_STATE_MEMORY_SLOTS
+        for shard_index in range(len(state.shard_buffer[buffer_index])):
+            committee_work = state.shard_buffer[buffer_index][shard_index]
+            if committee_work.status.selector == CONFIRMED_SHARD_DATA:
+                # Charge EIP 1559 fee
+                proposer = get_shard_proposer_index(state, slot, Shard(shard_index))
+                fee = (
+                    (state.shard_gasprice * candidate.commitment.length)
+                    // TARGET_SAMPLES_PER_BLOCK
+                )
+                decrease_balance(state, proposer, fee)
+
+                # Track updated gas price
+                new_gasprice = compute_updated_gasprice(
+                    new_gasprice,
+                    candidate.commitment.length,
+                    adjustment_quotient,
+                )
     state.shard_gasprice = new_gasprice
 ```
 
+#### `reset_pending_shard_work`
+
 ```python
-def reset_pending_headers(state: BeaconState) -> None:
-    state.previous_epoch_pending_shard_headers = state.current_epoch_pending_shard_headers
-    state.current_epoch_pending_shard_headers = []
-    # Add dummy "empty" PendingShardHeader (default vote for if no shard header available)
+def reset_pending_shard_work(state: BeaconState) -> None:
+    # Add dummy "empty" PendingShardHeader (default vote if no shard header is available)
     next_epoch = get_current_epoch(state) + 1
     next_epoch_start_slot = compute_start_slot_at_epoch(next_epoch)
     committees_per_slot = get_committee_count_per_slot(state, next_epoch)
+    active_shards = get_active_shard_count(state, next_epoch)
+
     for slot in range(next_epoch_start_slot, next_epoch_start_slot + SLOTS_PER_EPOCH):
-        for index in range(committees_per_slot):
-            committee_index = CommitteeIndex(index)
-            shard = compute_shard_from_committee_index(state, slot, committee_index)
+        buffer_index = slot % SHARD_STATE_MEMORY_SLOTS
+        
+        # Reset the shard work tracking
+        state.shard_buffer[buffer_index] = [ShardWork() for _ in range(active_shards)]
+
+        start_shard = get_start_shard(state, slot)
+        for committee_index in range(committees_per_slot):
```

Comment:
Might it be a good idea to just reset _all_ the ShardWork, even if those shards are not active? I worry that having some random shards in some cases stay past `SHARD_STATE_MEMORY_SLOTS` won't be very useful because it's unreliable, and it could risk causing bugs from a record corresponding to an older slot being mistaken for a record referring to a newer slot.

Reply:
Do you mean the 778 line? There the list for the given slot is overwritten as a whole, it's not just the `0...len(active_shards)` part that gets overwritten, that's just what the new list is filled with. So it can change shard count without problem.

Regarding the outer layer with `buffer_index` itself; it's a ring buffer, so we have to reset slot by slot. Currently it keeps the latest 256 slots in memory, and resets the last 32 every epoch. Resetting slot-by-slot may be better, but I'm not sure if we want to change the `process_slots` processing.

Reply:
Aaah I missed the 778 line; I thought only the active shards were being reset. I'm satisfied now.